keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+
+![Network Proxy Simple Cluster](NetworkProxySimpleCluster.png)
+The diagram shows API Server’s outgoing traffic flow. 
+The user (in blue box), master network (in purple cloud) and 
+a cluster network (in green cloud) are represented.  
+
+The user (blue) initiates communication to the KAS. 
+The KAS then initiates connections to other components. 
+It could be node/pod/service in cluster networks (red dotted arrow to green cloud), 
+or etcd for storage in the same master network (blue arrow) or mutate the request 
+based on an admission web-hook (red dotted arrow to purple cloud). 
+The KAS handles these cases based on NetworkContext based traffic routing. 
+The connectivity proxy should be able to do routing solely based on IP. 
+The proxy should not require the NetworkContext. This means the service CIDR, 
+node CIDR and pod CIDR of each cluster network cannot overlap. 
+
+### Network Context
+
+The minimal NetworkContext looks like the following struct in golang:
+
+```go
+type NetworkContext struct {
+        // NetworkName is the unique name of the network 
+        // to select to route the traffic to. 
+        NetworkName string
+}
+```
+
+NetworkName specifies the network to route traffic to. 
+The KAS starts with a list of registered network names and 
+the corresponding Connectivity services. So the KAS knows where to 
+proxy the traffic to, otherwise it return an “Unknown network” error. 
+
+The KAS starts with a proxy configuration like the below example. 
+The example specifies 4 networks. "direct" specifies the KAS talking directly 
+on the local network (no proxy). "master" specifies the KAS talks to a proxy
+listening at 1.2.3.4:5678. "cluster" specifies the KAS talk to a proxy 
+listening at 1.2.3.5:5679. While these are represented as resources 
+they are not intended to be loaded dynamically. 
+The KAS loads them as configuration at start time.
+
+```yaml
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: direct
+  connection:
+    type: direct
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Master
+  connection:
+    type: grpc
+    url: grpc://1.2.3.4:5678
+    caBundle: file1.ca
+---
+apiVersion: connectivityservice.k8s.io/v1alpha1
+kind: ConnectivityServiceConfiguration
+connectionService:
+  name: Cluster
+  connection:
+    type: grpc
+    url: grpc://1.2.3.5:5679
+    caBundle: file2.ca
+
+```
+
+NetworkContext could be extended to contain more contextual information.
+This would allow smarter routing based on the k8s object KAS is processing 
+or which user/tenant tries to initiate the request, etc.
+
+### Proxy gRPC definition
+
+```grpc
+syntax = "proto3";
+
+service ProxyService {
+    // Proxy connects to a remote address by stream id, and establish
+    // a bi-directional stream of packet exchange.
+    rpc Proxy(stream Packet) returns (stream Packet) {}
```

Comment:
Why wouldn't a simpler interface work? e.g.
```
rpc Proxy(stream Bytes) (stream Packet) {}
```
With parameters like IP passed in headers?

Reply:
Maybe it could. I thought stream as a way of a reliable or unreliable transport, like IP network. then I thought of packet as a TCP/UDP packet, where the functionality is self-contained, not relying on anything else.

Reply:
It depends on if you ever want to support more than the https://godoc.org/net#Conn interface. If net.Conn (which supports UDP connections) is sufficient then this is a massive simplification.